packages/components/src/custom-select/types.ts

```diff
@@ -0,0 +1,38 @@
+// TO DO
+export type CustomSelectProps = {
+	/**
+	 * The children elements. This should be composed
+	 * of CustomSelect.Item components.
+	 */
+	children: React.ReactNode;
+	/**
+	 * An optional default value for the control.
+	 * If left undefined, the first non-disabled
+	 * item will be used
+	 */
+	defaultValue?: any;
+	/**
+	 * Label for the control.
+	 */
+	label?: string;
+	/**
+	 * A function that receives the new value of the input.
+	 */
+	onChange?: Function;
+	/**
+	 * The size of the select.
+	 */
+	size?: 'default' | 'large';
```

Comment:
We'll need to decide / have design input on what size the large would be. Right now, it's set to auto to accommodate the height but I'm sure that could lead to issues if the content was ridiculously big.

Reply:
Yeah, this definitely needs design feedback. We should probably also discuss whether having "rich" previous for the  selected value is something we should do at all, since the legacy `CustomSelectControl` doesn't currently support it.

Reply:
Will we want to get design feedback before merging this PR? Or as a follow-up?

Reply:
Sounds good, thanks for pinging them! I'll add a summary below: 

### Sizing

This new version of `CustomSelectControl` can render children, allowing for more customization. [Ariakit has a great example](https://ariakit.org/examples/select-item-custom) of a potential use case, a more visual account selection: 

<img width="351" alt="Screenshot 2023-11-15 at 1 43 55 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/a7327479-fd63-403a-849b-8107c85c4f90">

The 40px 'large' height would be too constrained for the select input preview: 

<img width="358" alt="Screenshot 2023-11-08 at 4 43 13 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/bf0518b7-1874-434e-a124-ca2820402759">

### Questions for Design

1. [As Marco pointed out](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1389302751), do we want to support rich 'previews' like in the example above? 
2. If so, how do we want to handle the sizes for rendered children?

Reply:
> do we want to support rich 'previews' like in the example above?

I would say yes it would be nice to have that level of flexibility. I'd expect that if a child is provided height is set to auto otherwise we default to the standard sizing of other inputs (40, 32). 

It's also worth connecting the dots with the [data views work](https://github.com/WordPress/gutenberg/issues/55083) as that has some filtering requirements that overlap somewhat with variations of select and dropdown menus.

Reply:
>I would say yes it would be nice to have that level of flexibility. I'd expect that if a child is provided height is set to auto otherwise we default to the standard sizing of other inputs (40, 32).

Thanks for the input! From my understanding of the [design specs](https://github.com/WordPress/gutenberg/issues/46734), the default would be `40px` for a select. So if rendered children will have a height of auto, then maybe we don't need a `size` prop? 🤔 Unless there is a plan for a `32px` 'compact/small' size as well? 

<img src="https://user-images.githubusercontent.com/1204802/209149927-4279c3e3-f085-42f1-aeed-dfed42627453.png"/>

>It's also worth connecting the dots with the https://github.com/WordPress/gutenberg/issues/55083 as that has some filtering requirements that overlap somewhat with variations of select and dropdown menus.

Thank you for sharing, I will take a look!

Reply:
> Unless there is a plan for a 32px 'compact/small' size as well?

I think there is! Moreover current version of `CustomSelectControl` already has a `size` prop accepting `small`, `default` and `__unstable-large` values:

- `small` is `24px` (I guess we can keep it around for now)
- `default` is either `32px` or `40px` depending on the `__next40pxDefaultSize` prop (long-term, default will be `40px)
- `__unstable-large` is `40px`, but it will be removed once the `__next40pxDefaultSize` is finished

We should probably replicate (the same `size` scheme in this new version; we could add, at most, a new `compact` size (`32px`) — this doesn't need to happen in this PR, it can also happen in a dedicated follow up.

> So if rendered children will have a height of auto

The way this could work is that, when the `renderSelectedValue` prop is defined:
 - we set a height of `auto`
 - optionally, the `height` associated to the `size` could become a `min-height` instead

Reply:
If we add `compact` size in a follow up and `__unstable-large` will be removed in the future, then in this PR, we would just have `small` and `default` for now? Based on my understanding that this is experimental and not exported, so we won't need to add `__unstable-large` and `__next40pxDefaultSize`.

Reply:
I'm ok with having just `small` and `default` for now, and adding `compact` at a later point.

Regarding `__unstable-large` and `__next40pxDefaultSize`, we may need to add at some point if the legacy CustomSelectControl will still support them when we swap its implementation with the new implementation.